### PR TITLE
2.2.6 Use batches and direct queries to fix sales address upgrade

### PR DESCRIPTION
**Pull request by rikwillems**
### Description (*)
Magento Commerce compatible fix for sales address upgrade script.

### Fixed Issues (if relevant)

1. magento/magento2#16570: enhance performance on large catalog
[(comment)](https://github.com/magento/magento2/pull/16570#issuecomment-429830198)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=18945) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=18945) before we can accept your contribution.<br/><hr/>**** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

**Comment by magento-engcom-team** [comment]
Hi @rikwillems. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me $VERSION instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by ihor-sviziev** [comment]
Hi @rikwillems,
Please sign CLA, otherwise we can't accept your PR

File: app/code/Magento/Sales/Setup/UpgradeData.php [hunk 1]
```diff
@@ -178,27 +179,77 @@ private function convertSerializedDataToJson($setupVersion, SalesSetup $salesSet
...
-            $addressTable
-        );
-        $setup->getConnection()->query($updateOrderAddress);
+            ->where('quote.entity_id in (?)', array_keys($orderAddresses))
+            ->where('address_type = ?', $address_type)
+        ;
+
+        $quoteAddresses = $quoteConnection->fetchAssoc($query);
+
+        $sql = "update `{$setup->getTable('sales_order_address')}` set `quote_address_id` = :quoteAddressId where `entity_id` = :orderAddressId";
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Setup/UpgradeData.php]: Looks like there is mistake:
```suggestion
        $sql = "update `{$setup->getTable('sales_order_address', 'sales')}` set `quote_address_id` = :quoteAddressId where `entity_id` = :orderAddressId";
```

File: app/code/Magento/Sales/Setup/UpgradeData.php [hunk 2]
```diff
@@ -178,27 +179,77 @@ private function convertSerializedDataToJson($setupVersion, SalesSetup $salesSet
      */
     public function fillQuoteAddressIdInSalesOrderAddress(ModuleDataSetupInterface $setup)
     {
-        $addressTable = $setup->getTable('sales_order_address');
-        $updateOrderAddress = $setup->getConnection()
+        $this->fillQuoteAddressIdInSalesOrderAddressByType($setup,Address::TYPE_SHIPPING);
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Setup/UpgradeData.php]:
```suggestion
        $this->fillQuoteAddressIdInSalesOrderAddressByType($setup, Address::TYPE_SHIPPING);
```

File: app/code/Magento/Sales/Setup/UpgradeData.php [hunk 3]
```diff
@@ -178,27 +179,77 @@ private function convertSerializedDataToJson($setupVersion, SalesSetup $salesSet
...
+        $this->fillQuoteAddressIdInSalesOrderAddressByType($setup,Address::TYPE_BILLING);
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Setup/UpgradeData.php]:
```suggestion
        $this->fillQuoteAddressIdInSalesOrderAddressByType($setup, Address::TYPE_BILLING);
```

**Comment by rikwillems** [comment]
@ihor-sviziev I did that and see it reflected in the ticket now.

**Comment by ihor-sviziev** [comment]
Hi @rikwillems,
Here is some failures in static and integration tests, looks like there is some issue in your implementation. Could you looks on it?

Your implementation actually not so easy. Could you describe how your solution works to be sure we're on the same page?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by rikwillems** [comment]
@ihor-sviziev 
It basically does the same thing as the other implementations. It works on raw queries but does not join between quote and sales. Also it works in batches to precent memory limits.

**Comment by ihor-sviziev** [comment]
Hi @rikwillems,
Your first commit https://github.com/magento/magento2/pull/18945/commits/8b5714790f1d7e85f78af69c17d24d1afebeb98c has different commiter email, as result - CLA showing that you didn't signed. Could you add that email to your github account OR just squash all your changes into single commit and force-push it?

**Comment by rikwillems** [comment]
I added PR #19098

**Comment by ihor-sviziev** [comment]
@rikwillems  thank you!
I'll close this PR as it is duplicates https://github.com/magento/magento2/pull/19098

**Pull request closed without merging**